### PR TITLE
Evolvable Hivemind

### PR DESCRIPTION
**Pull request by Toxici11i**
## About The Pull Request
This allows hiveminds to well, as the title suggests, be an evolvable cast (from drone)

## Why It's Good For The Game
Currently hiveminds are only creatable from shrikes/queens, which require time, and plasma, to do so. Along with them, yknow, remembering hiveminds exist. This allows any drone to become a hivemind (creating the core on their position) and in general is an improvement.
There are also various other QOL changes to hiveminds here, namely being a larger weed jump radius (if a xeno is within 6 tiles of weeds, you can jump to them) and loudvoice for easier communication

image
image

## Changelog
:cl:
del: removed hivemind cores automatically spawning a hivemind (to prevent queens from spawning another)
add: the ability for drones to evolve into hiveminds
del: queens and shrikes being able to manually spawn them
tweak: hiveminds now have loudvoice
tweak: hiveminds no longer display their name in xeno hivemind, instead are just 'The hivemind'
/:cl:

### CODE REVIEW

**Comment by Toxici11i** [comment]
Travis compile failure is due to me having to manually include the dependencies for the code to compile, one second.

**Comment by Toxici11i** [comment]
This should be ready for merge

File: code/game/objects/effects/weeds.dm [hunk 1]
```diff
@@ -179,7 +179,7 @@
 
 /obj/effect/alien/weeds/node/hivemindcore/Initialize(mapload, obj/effect/alien/weeds/node/node)
 	. = ..()
-	parent = new(loc)
+	//parent = new(loc) //deprecated due to changes
```
Comment (Rohesie) [inline code/game/objects/effects/weeds.dm]: Remove old commented code once you are done.

File: code/modules/mob/living/carbon/xenomorph/abilities.dm [hunk 2]
```diff
@@ -849,6 +849,7 @@
 // ***************************************
 // *********** Spawn hivemind
 // ***************************************
+/*
```
Comment (Rohesie) [inline code/modules/mob/living/carbon/xenomorph/abilities.dm]: Same, remove old commented code once you are done.

File: code/modules/mob/living/carbon/xenomorph/castes/hivemind/hivemind.dm [hunk 3]
```diff
@@ -69,7 +69,7 @@
 		var/mob/living/carbon/xenomorph/xeno = locate(href_list["hivemind_jump"])
 		if(!istype(xeno))
 			return
-		var/obj/effect/alien/weeds/nearby_weed = locate() in range("3x3", get_turf(xeno))
+		var/obj/effect/alien/weeds/nearby_weed = locate() in range("6x6", get_turf(xeno)) //increased
```
Comment (Rohesie) [inline code/modules/mob/living/carbon/xenomorph/castes/hivemind/hivemind.dm]: Consider that this kind of comment is not really informative for future coders, as it has no context.

File: code/modules/mob/living/carbon/xenomorph/evolution.dm [hunk 4]
```diff
@@ -196,6 +196,15 @@
 			to_chat(src, "<span class='warning'>Something in this place is interfering with our link to Queen Mother. We are unable to evolve to a psychic caste here!</span>")
 			return
 
+	else if(new_caste_type == /mob/living/carbon/xenomorph/hivemind) //Special case for dealing with hiveminds - this may be subject to heavy change, such as multiple hiveminds potentially being an option
+		if(length(hive.xenos_by_typepath[/mob/living/carbon/xenomorph/hivemind]))
+			to_chat(src, "<span class='warning'>There cannot be two manfiestations of the hiveminds will at once..</span>")
```
Comment (Rohesie) [inline code/modules/mob/living/carbon/xenomorph/evolution.dm]:
```suggestion
			to_chat(src, "<span class='warning'>There cannot be two manfiestations of the hivemind's will at once..</span>")
```

File: code/modules/mob/living/carbon/xenomorph/evolution.dm [hunk 5]
```diff
@@ -280,8 +289,18 @@
...
+			var/obj/effect/alien/weeds/node/hivemindcore/newcore = new /obj/effect/alien/weeds/node/hivemindcore(new_xeno.loc)
+			newcore.parent = newmind
+			newmind.core = newcore
 	else
 		new_xeno.key = key
+		if(new_caste_type == /mob/living/carbon/xenomorph/hivemind)
+			var/mob/living/carbon/xenomorph/hivemind/newmind = new_xeno
+			var/obj/effect/alien/weeds/node/hivemindcore/newcore = new /obj/effect/alien/weeds/node/hivemindcore(new_xeno.loc)
+			newcore.parent = newmind
+			newmind.core = newcore
```
Comment (Rohesie) [inline code/modules/mob/living/carbon/xenomorph/evolution.dm]: This is duplicated code.
Add a `post_evolve()` redefined proc or something.

File: code/modules/mob/living/carbon/xenomorph/castes/hivemind/hivemind.dm (hunk 3 above)
Comment (Toxici11i) [reply code/modules/mob/living/carbon/xenomorph/castes/hivemind/hivemind.dm]: fair

File: code/modules/mob/living/carbon/xenomorph/evolution.dm (hunk 5 above)
Comment (Rohesie) [reply code/modules/mob/living/carbon/xenomorph/evolution.dm]: Actually redefining it won't work.
The ideal way would be doing a xeno caste `on_creation()` proc or the likes, but if you don't feel like refactoring this just add this code after the `if-else`, since it's happening anyway on the same format.

File: code/modules/mob/living/carbon/xenomorph/evolution.dm (hunk 4 above)
Comment (Rohesie) [reply code/modules/mob/living/carbon/xenomorph/evolution.dm]: `hiveminds` is not the same as `hivemind's`

File: code/modules/mob/living/carbon/xenomorph/castes/drone/castedatum_drone.dm [hunk 6]
```diff
@@ -28,7 +28,7 @@
 	evolution_threshold = 80
 	upgrade_threshold = 50
 
-	evolves_to = list(/mob/living/carbon/xenomorph/shrike, /mob/living/carbon/xenomorph/carrier, /mob/living/carbon/xenomorph/hivelord)
+	evolves_to = list(/mob/living/carbon/xenomorph/shrike, /mob/living/carbon/xenomorph/carrier, /mob/living/carbon/xenomorph/hivelord,/mob/living/carbon/xenomorph/hivemind)
```
Comment (psykzz) [inline code/modules/mob/living/carbon/xenomorph/castes/drone/castedatum_drone.dm]:
```suggestion
	evolves_to = list(/mob/living/carbon/xenomorph/shrike, /mob/living/carbon/xenomorph/carrier, /mob/living/carbon/xenomorph/hivelord, /mob/living/carbon/xenomorph/hivemind)
```

File: code/modules/mob/living/carbon/xenomorph/castes/hivemind/hivemind.dm [hunk 7]
```diff
@@ -30,11 +30,12 @@
...
 
-	var/obj/effect/alien/weeds/node/hivemindcore/core
+	var/obj/effect/alien/hivemindcore/core
 
-/mob/living/carbon/xenomorph/hivemind/Initialize(mapload, obj/effect/alien/weeds/node/hivemindcore/hmcore)
+/mob/living/carbon/xenomorph/hivemind/Initialize(mapload)
 	. = ..()
-	core = hmcore
+	core = new /obj/effect/alien/hivemindcore(loc, src)
+	new /obj/effect/alien/weeds/node/strong(loc)
```
Comment (Rohesie) [inline code/modules/mob/living/carbon/xenomorph/castes/hivemind/hivemind.dm]: Weeds first, hivemind after.

File: code/modules/mob/living/carbon/xenomorph/evolution.dm [hunk 8]
```diff
@@ -249,6 +257,10 @@
 		if(length(hive.xenos_by_typepath[/mob/living/carbon/xenomorph/shrike]))
 			to_chat(src, "<span class='warning'>There already is a Shrike.</span>")
 			return
+	else if(new_caste_type == /mob/living/carbon/xenomorph/hivemind) //Special case for dealing with hiveminds - this may be subject to heavy change, such as multiple hiveminds potentially being an option
+		if(length(hive.xenos_by_typepath[/mob/living/carbon/xenomorph/hivemind]))
+			to_chat(src, "<span class='warning'>There cannot be two manfiestations of the hiveminds will at once..</span>")
```
Comment (Rohesie) [inline code/modules/mob/living/carbon/xenomorph/evolution.dm]:
```suggestion
			to_chat(src, "<span class='warning'>There cannot be two manifestations of the hivemind's will at once.</span>")
```

**Comment by Rohesie** [comment]
Code itself is fine. It needs balance and game design approval however.

File: code/modules/mob/living/carbon/xenomorph/evolution.dm [hunk 9]
```diff
@@ -196,6 +195,15 @@
...
 
+	else if(new_caste_type == /mob/living/carbon/xenomorph/hivemind) //Special case for dealing with hiveminds - this may be subject to heavy change, such as multiple hiveminds potentially being an option
+		if(length(hive.xenos_by_typepath[/mob/living/carbon/xenomorph/hivemind]))
+			to_chat(src, "<span class='warning'>There cannot be two manifestations of the hivemind's will at once..</span>")
+			return
+
+		if(isxenoresearcharea(get_area(src)))
+			to_chat(src, "<span class='warning'>Something in this place is interfering with our link to the Hivemind. We are unable to evolve to be its manifestation!</span>")
+			return
+
```
Comment (psykzz) [inline code/modules/mob/living/carbon/xenomorph/evolution.dm]: since a hivemind is like a structure, you should do the same structure checks

File: code/modules/mob/living/carbon/xenomorph/castes/hivemind/hivemind.dm [hunk 10]
```diff
@@ -69,7 +70,7 @@
 		var/mob/living/carbon/xenomorph/xeno = locate(href_list["hivemind_jump"])
 		if(!istype(xeno))
 			return
-		var/obj/effect/alien/weeds/nearby_weed = locate() in range("3x3", get_turf(xeno))
+		var/obj/effect/alien/weeds/nearby_weed = locate() in range("6x6", get_turf(xeno)) //increased
```
Comment (psykzz) [inline code/modules/mob/living/carbon/xenomorph/castes/hivemind/hivemind.dm]: Wasa there any reason this was changed in this PR? This is unrelated to this PR.

Comment (Toxici11i) [reply code/modules/mob/living/carbon/xenomorph/castes/hivemind/hivemind.dm]: @psykzz yes actually, the range you could find xenos in (jumping to) was unreliable in a 3x3 zone, even though you could then jump to them if you could manually go to weeds, and then doubleclick them

File: code/modules/mob/living/carbon/xenomorph/evolution.dm (hunk 9 above)
Comment (Toxici11i) [reply code/modules/mob/living/carbon/xenomorph/evolution.dm]: I dont see the point of doing any structure checks on hiveminds due to the fact you need to A) be a drone before evolving and B) evolving in a place where the structure check for nodes fails will actually be to your detriment, since you couldnt build defences around your core.

Comment (psykzz) [reply code/modules/mob/living/carbon/xenomorph/evolution.dm]: There are a number of things that would be beneficial to a hivemind but would fail, for example building a hivemind on top a carrier trap or even a resin bed.

Comment (Toxici11i) [reply code/modules/mob/living/carbon/xenomorph/evolution.dm]: or... just placed in fog where they are immortal

Comment (Toxici11i) [reply code/modules/mob/living/carbon/xenomorph/evolution.dm]: Point is, theres 0 concievable advantage of placing hivemind on normally unplacable tiles, since you could just do the same thing in fog (which you currently can).

Comment (psykzz) [reply code/modules/mob/living/carbon/xenomorph/evolution.dm]: Thats besides the point, structure creation for aliens currently run the check_alien_construction.

Whether or not it makes sense for the user, for the codebase it allows us to control structure creation from a single place.

I agree there are a lot of things that can be improved regarding the hivemind and im not opposed to them.

This however is seperate, and shouldnt be seen as the same.

**Comment by Toxici11i** [comment]
Latest change stopped xenos from bumping into hiveminds (IT SEEMS) along with fixing marines being able to hit hiveminds, or see their position entirely even with rightclick.

**Comment by Kuro-ai** [comment]
Hivemind is a broken mess,tm this already

**Comment by Pariah919** [comment]
as the saltiest man on earth
i approve this

**Comment by Toxici11i** [comment]
reverting back to previous iteration of assigning hiveminds to cores, as its less bugggy.

File: code/modules/mob/living/carbon/xenomorph/evolution.dm [hunk 11]
```diff
@@ -196,6 +195,15 @@
 			to_chat(src, "<span class='warning'>Something in this place is interfering with our link to Queen Mother. We are unable to evolve to a psychic caste here!</span>")
 			return
 
+	else if(new_caste_type == /mob/living/carbon/xenomorph/hivemind) //Special case for dealing with hiveminds - this may be subject to heavy change, such as multiple hiveminds potentially being an option
```
Comment (spookydonut) [inline code/modules/mob/living/carbon/xenomorph/evolution.dm]: trying to special case evolution like this with a mob type is really hacky and a hard no.

Comment (Toxici11i) [reply code/modules/mob/living/carbon/xenomorph/evolution.dm]: @spookydonut read above, literally the exact same method as checking for multiple queens and shrikes.

**Comment by Toxici11i** [comment]
> just leave it how it is except delete the xeno using the create hivemind ability, and make them the hivemind instead of offering the hivemind

you obviously dont understand how hiveminds were created, and why that wouldnt work easily.
You also dont seem to understand that only shrikes/queens could spawn hiveminds.
Nor have you read previous code relating to how single-evolve castes were handled.

Stop it, get some help.

File: code/modules/mob/living/carbon/xenomorph/evolution.dm (hunk 11 above)
Comment (Toxici11i) [reply code/modules/mob/living/carbon/xenomorph/evolution.dm]: `	if(new_caste_type == /mob/living/carbon/xenomorph/queen)
		if(hive.living_xeno_queen) //Do another check after the tick.
			to_chat(src, "<span class='warning'>There already is a Queen.</span>")
			return
	else if(new_caste_type == /mob/living/carbon/xenomorph/shrike)
		if(length(hive.xenos_by_typepath[/mob/living/carbon/xenomorph/shrike]))
			to_chat(src, "<span class='warning'>There already is a Shrike.</span>")
			return
	else if(new_caste_type == /mob/living/carbon/xenomorph/hivemind) //Special case for dealing with hiveminds - this may be subject to heavy change, such as multiple hiveminds potentially being an option
		if(length(hive.xenos_by_typepath[/mob/living/carbon/xenomorph/hivemind]))
			to_chat(src, "<span class='warning'>There cannot be two manifestations of the hivemind's will at once.</span>")
			return`

**Comment by spookydonut** [comment]
> 
> 
> > just leave it how it is except delete the xeno using the create hivemind ability, and make them the hivemind instead of offering the hivemind
> 
> you obviously dont understand how hiveminds were created, and why that wouldnt work easily.
> You also dont seem to understand that only shrikes/queens could spawn hiveminds.
> Nor have you read previous code relating to how single-evolve castes were handled.
> 
> Stop it, get some help.

I wrote most of this code, I don't know what you seek to gain by being hostile to maintainers

**Comment by Toxici11i** [comment]
le classic tg maintainer.

**Comment by Toxici11i** [comment]
> > > just leave it how it is except delete the xeno using the create hivemind ability, and make them the hivemind instead of offering the hivemind
> > 
> > 
> > you obviously dont understand how hiveminds were created, and why that wouldnt work easily.
> > You also dont seem to understand that only shrikes/queens could spawn hiveminds.
> > Nor have you read previous code relating to how single-evolve castes were handled.
> > Stop it, get some help.
> 
> I wrote most of this code, I don't know what you seek to gain by being hostile to maintainers

If you wrote most of it, why are you challenging a valid point about the code?. Using the standards set in evolution code ( i was told to, by another maintainer, add an additional check using the same code) and then seeking a complete change to how the ability itself works, and suggesting a bad idea (inherently deleting a shrike/queen for a hivemind), you need to reevaluate why you are a maintainer.
Please, get some help. Im reopening this since your both banned from the discord, and unwilling to actually say why using previously approved code is invalid.
